src/lib/libpz/pzip.3
.TH PZIP 3 "1998-08-11"
.SH NAME \" @(#)pzip.3 (gsf@research.att.com) 1998-08-11
pzip \- fixed record partition compress/decompress library
.SH SYNOPSIS
.fp 5 CW
.de L		\" literal font
.ft 5
.if !\\$1 \&\\$1 \\$2 \\$3 \\$4 \\$5 \\$6 \f1
..
.de LR
.}S 5 1 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de RL
.}S 1 5 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de LI
.}S 5 2 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de IL
.}S 2 5 \& "\\$1" "\\$2" "\\$3" "\\$4" "\\$5" "\\$6"
..
.de EX		\" start example
.ta 1i 2i 3i 4i 5i 6i
.PP
.RS 
.PD 0
.ft 5
.nf
..
.de EE		\" end example
.fi
.ft
.PD
.RE
.PP
..
.de Tp
.fl
.ne 3
.TP
..
.de Ss
.fl
.ne 3
.SS "\\$1"
..
.ta 1.0i 2.0i 3.0i 4.0i 5.0i
.nf
.ft 5
#include   <pzip.h>
.ft 1
.fi
.Ss "DATA TYPES"
.nf
.ft 5
Pz_t;
Pzpart_t;
Pzdisc_t;

int        (*Pzerror_f)(Pz_t*, Pzdisc_t*, int, const char*, ...);
int        (*Pzcheck_f)(Pz_t*, unsigned char*, Pzdisc_t*);
.ft 1
.fi
.Ss "CONSTANTS"
.nf
.ft 5
PZ_VERSION

PZ_WINDOW

PZ_MAGIC_1
PZ_MAGIC_2

PZ_TEST
.ft 1
.fi
.Ss "FLAGS"
.nf
.ft 5
PZ_READ
PZ_WRITE
PZ_FORCE
PZ_STAT
PZ_STREAM
PZ_CRC
PZ_DUMP
PZ_VERBOSE
PZ_TEST_1
PZ_TEST_2
PZ_TEST_3
PZ_TEST_4

SFPZ_VERIFY
SFPZ_CRC
.ft 1
.fi
.Ss "OPENING/CLOSING STREAMS"
.nf
.ft 5
Pz_t*      pzopen(Pzdisc_t* disc, const char* path, int flags);
int        pzclose(Pz_t* pz);
.ft 1
.fi
.Ss "INPUT/OUTPUT OPERATIONS"
.nf
.ft 5
int        pzdeflate(Pz_t* pz, Sfio_t* out);
int        pzinflate(Pz_t* pz, Sfio_t* out);

int        pzgethdr(Pz_t* pz);
int        pzputhdr(Pz_t* pz, Sfio_t* out);

.ft 1
.fi
.Ss "STREAM INFORMATION"
.nf
.ft 5
Pzpart_t*  pzgetpart(Pz_t* pz, const char* name);
Pzpart_t*  pzsetpart(Pz_t* pz, Pzpart_t* part);

void       pzdump(Pz_t* pz, Sfio_t* out);
.ft 1
.fi
.Ss "SFIO DECOMPRESS DISCIPLINE"
.nf
.ft 5
int        sfdcpzip(Sfio_t* stream, const char* options, int flags);
.ft 1
.fi
.SH DESCRIPTION
.PP
.I pzip
is a library of functions that support compression/decompression
of data files of fixed length rows (records) and columns (fields).
Each
.I pzip
stream represents a file of compressed or decompressed data
that may be decompressed or compressed to an
.IR sfio (3)
output file stream.
.PP
.I pzip
format performs better than
.I gzip
in space/time on data that has many (> 50%) columns
that change at a low rate
(columns with a low rate of change are low frequency;
columns with a high rate of change are high frequency.)
.PP
The
.I pzip
compress format is itself
.I gzipped
using the
.IR sfdcgzip ()
.I sfio
discipline.
Decompressed data is reorganized according to the user-specified
.I partition-file
(see
.L Pzdisc_t.partition
below) before being passed to
.IR gzip .
Low frequency columns are run-length encoded and high frequency column groups
are transposed to column-major order.
The
.I gzip
tables are flushed, via the
.IR sfsync ()
discipline function, between each column partition group.
This has a positive space/time effect on the
.I gzip
string match and huffman tables.
.PP
.I pzip
format files self-idenfify by encoding partition information in a header.
.PP

.Ss "DATA TYPES"
.PP
.Ss "  Pz_t"
This is the stream handle returned by
.IR pzopen .
None of the fields should be changed by the application.
The fields are:
.Tp
.L "const char* id"
The library identification string used by the
.I libast
.I errorf
function to identify the source of error and warning messages.
.Tp
.L "Pzdisc_t* disc"
A pointer to the user discipline structure from
.IR pzopen .
.Tp
.L "unsigned int flags"
The inclusive-or of the flags described below.
.Tp
.L "int major"
The
.I pzip
file format major number.
For
.L PZ_READ
it is the the value when the file was compressed,
otherwise it is the major number of the current implementation.
In general the file header and implementation major numbers must match.
.Tp
.L "int minor"
The
.I pzip
file format minor number.
For
.L PZ_READ
it is the the value when the file was compressed,
otherwise it is the minor number of the current implementation.
In general, all implementations and data files with the same
file header major number are compatible.
The minor allows the implementation to make runtime adjustments.
.Tp
.L "size_t win"
The high frequency window size.
This is the adjusted value that is divisible by both the row size and
number of high frequency columns.
.Tp
.L "const char* path"
The input file pathname.
.Tp
.L "Sfio_t* io"
The input file
.I sfio
stream.
.Tp
.L "Vmalloc_t* vm"
The
.IR vmalloc (3)
region handle for the entires stream.
All allocations associated with the stream, except for
.I sfio
intiated allocations, are done in this region.
The user may also allocate and free individual chunks of memory from
this region, but must not call
.IR vmclear ().
The region is freed by
.IR pzclose ().
.Tp
.L "size_t npart"
The number of partitions.
.Tp
.L "Pzpart_t* part"
The current active partition.
.Tp
.L "Pzpart_t* parts"
A table of all partitions.
.Tp
.L "unsigned char* buf"
The high frequency window buffer with
.L win
elements.
.Tp
.L "unsigned char* wrk"
The PZ_WRITE high frequency window buffer with
.L win
elements.
.Tp
.L "unsigned char* pat"
The low frequency pattern buffer with
.L row
elements.
.PP

.Ss "  Pzpart_t"
.PP
.L Pzpart_t
defines one partition.
The fields are:
.Tp
.L "char* name"
The partition name.
.Tp
.L "int index"
The partition index.
May be used to access a partition given its index:
.L "Pz_t.parts[Pzpart_t.index-1]."
.Tp
.L "size_t row"
The partition fixed row size.
.Tp
.L "size_t col"
The number of rows that can fit into the high frequency window column buffer.
.Tp
.L "size_t* map"
An array with
.L nmap
elements that lists the high frequency column
indexes in order.
.Tp
.L "size_t* grp"
An array with
.L ngrp
elements that lists the sizes of each
high frequency column partition group in the same order as
.LR map .
.Tp
.L "size_t nmap"
The number of elements in
.LR map .
.Tp
.L "size_t ngrp"
The number of elements in
.LR grp .
.Tp
.L "unsigned char* low"
An array with
.L row
elements.
.LI low[ i ]
is
.L 1
if column
.I i
is low frequency,
otherwise it is
.L 0
(and column
.I i
is high frequency.)
.Tp
.L "int* value"
If there are no fixed-value columns the
.L value
is
.LR NULL .
Otherwise it is an array with
.L row
elements.
.LI value[ i ]
is non-negative if column
.I i
has a fixed value
(and
.LI value[ i ]
is the fixed column value).
.Tp
.L "size_t* fix"
An array with
.L nfix
elements that lists fixed value columns.
.Tp
.L "size_t nfix"
The number of elements in
.LR fix .
.PP

.Ss "  Pzdisc_t"
.PP
.L Pzdisc_t
defines a stream discipline structure to the
.IR pzopen ()
function.
The discipline fields are:
.Tp
.L "unsigned long version"
Must be initailized to
.LR PZ_VERSION .
.Tp
.L "const char* comment"
An optional string that is placed in the
.I pzip
output file header; this string can be viewed by the
.IR pzip (1)
command.
Ignored for
.L PZ_READ
and
.L PZ_STAT
streams.
.Tp
.L "const char* options"
An optional string of run-time options of the form
.IR name = value .
Currently only fixed value columns may be specified.
The syntax is
.IR begin [ -end ]= "'value'"
where
.I begin
is the beginning column offset (starting at 0),
.I end
is the ending column offset for an inclusive range,
and
.I value
is the fixed column value.
Decompress time is improved when high frequency columns are given fixed values.
.Tp
.L "const char* partition"
The name of the
.I partition-file
that contains a sequence of lines that
specifie the data row size and the high frequency
column partition groups.
This entry must be specified for
.L PZ_WRITE
and is ignored for
.L PZ_READ
streams.
Comments start with # and continue to the end of the line.
The first non-comment line specifies the row size.
The remaining lines operate on column offset ranges of the form:
.IR begin [ -end ]
where
.I begin
is the beginning column offset (starting at 0),
.I end
is the ending column offset for an inclusive range.
The operations are:
.RS
.Tp
.IR range " [ ... ]"
places all columns in the specified
.I range
list in the same high frequency partition group.
Each high frequency partition group is processed as a separate block by
.IR gzip .
.Tp
.I "range='value'"
specifies that each column in
.I range
has the fixed character value
.IR value .
C-style character escapes are valid for
.IR value .
.RE
.Tp
.L "const char* lib"
The library name used by
.IR pathfind (3)
to locate partition files.
The default is
.LR '"pzip"' ,
and the default partition file suffix is
.LR .prt .
.Tp
.L "size_t window"
Low frequency columns are processed one row at a time;
high frequency columns are processed across many rows at a time.
The space/time tradeoff is controlled by the number of
high frequency columns that can be processed in one step.
.L window
sets this limit.
The high frequency columns are transposed from row-major order to
column-major order, which may bring on inefficient paging behavior
on some systems when the window size is too large.
The default of 4M (4194304) provides reasonable behavior across
most paging implementations.
Note that compression requires two
.L window
buffers whereas decompression requires one.
.L window
is shortened to be divisible by both the row size and the number
of high frequency columns.
.Tp
.L "int (*errorf)(Pz_t* pz, Pzdisc_t* disc, int lev, const char*, fmt ...)"
An optional function that is called to emit error and warning messages.
It is most often set to the
.I libast
.IR errorf ():
.EX
disc.errorf = (Pzerror_f)errorf;
.EE
.Tp
.L "int (*eventf)(Pz_t* pz, int event, void* value, Pzdisc_t* disc)"
An optional function that is called when events occur during stream processing.
.L event
is set to the event and 
.L value
is an event specific value.
The events are:
.RS
.Tp
.L PZ_OPEN
Called just before
.IR pzopen ()
returns successfully.
.L value
is
.LR 0 .
A
.L -1
return value causes
.IR pzopen ()
to fail.
.Tp
.L PZ_CLOSE
Called before
.IR pzclose ()
releases any resources.
.L value
is
.LR 0 .
The return value is used as the 
.IR pzclose ()
return value.
.Tp
.L PZ_CHECK
Called as each row in a
.L PZ_WRITE
stream is processed.
.L value
is a pointer to the row data before compression, and
.L eventf
may modify the contents up to the row size.
The return value determines the disposition of the row:
.L -1
terminates all processing;
.L 0
ignores the row;
otherwise the row is processed as usual.
.RE

.PP
.Ss "CONSTANTS"
.PP
.Tp
.L PZ_VERSION
This is a macro value of type
.L "long int"
that defines
the current version number of the
.I pzip
library interface.
The form is a six digit date YYYYMMDD.
.Tp
.L PZ_WINDOW
The default window size (4Mb).
.Tp
.L PZ_MAGIC_1
The first character of the two character
.I pzip
header magic number.
.Tp
.L PZ_MAGIC_2
The second character of the two character
.I pzip
header magic number.
.PP

.Ss "BIT FLAGS"
.PP
A number of bit flags control stream operations.
They are set by the
.L flags
argument to
.IR pzopen ().
The flags are:
.Tp
.L PZ_READ
The input file is opened for decompression to the output file.
.Tp
.L PZ_WRITE
The input file is opened for compression to the output file.
The input file is opened for decompression.
.Tp
.L PZ_FORCE
For
.LR PZ_READ ,
if the input file is not in
.I pzip
format, then 
.IR pzinflate ()
operates in transparent mode.
If the input file is in
.I gzip
format then
.I gzip
inflate is applied.
Otherwise
.L PZ_READ
input files must be in
.I pzip
format.
.Tp
.L PZ_STAT
The input file must be in
.I pzip
format; the handle may be used
to retrieve header information, but
.IR pzinflate ()
is disabled.
.Tp
.L PZ_STREAM
The
.L path
argument to
.IR pzopen ()
is treated as an
.I sfio
.L SF_READ
stream.
This is a hack used by
.IR sfdcpzip ().
.Tp
.L PZ_CRC
Enables decompress crc checking.
crc checking is a perfomance wart in the otherwise respectable
.IR libz (3)
.I gzip
library implementation.
Decompression crc checking can increase decompression user time by as much
as a factor of 2.
.I pzip
uses a version of
.I libz
that disables decompression crc checking
and replaces it with a few sanity checks.
The
.I pzip
format also has its own checks.
.Tp
.L PZ_DUMP
Calls
.IR pzdump ()
just before a successful return from
.IR pzopen ().
.Tp
.L PZ_VERBOSE
Enables a verbose trace of internal actions.
.Tp
.L PZ_TEST_1
Enables the implementation defined test #1.
.LR PZ_TEST_2 ,
.LR PZ_TEST_3 ,
and
.L PZ_TEST_4
also provided.
.PP

.Ss "OPENING/CLOSING STREAMS"

.PP
.nf
.ft 5
Pz_t*      pzopen(Pzdisc_t* disc, const char* path, int flags);
.ft 1
.fi
.PP
This function opens a stream on
.LR file .
It returns a new stream handle on success and
.L NULL
on error.
.L disc
and
.L flags
are described above.
If
.L flags
contains
.L PZ_READ
then
.IR pzinflate ()
may be called to decompress
.LR file ,
otherwise if
.L flags
contains
.L PZ_WRITE
.IR pzdeflate ()
may be called to compress
.LR file .

.PP
.nf
.ft 5
int        pzclose(Pz_t* pz);
.ft 1
.fi
This functions close the stream handle
.L pz
returned by a previous call to
.IR pzopen ().
It returns
.L 0
on success and
.L -1
on error.
All resources allocated on behalf of the stream are released.

.PP
.Ss "INPUT/OUTPUT OPERATIONS"

.PP
.nf
.ft 5
int        pzdeflate(Pz_t* pz, Sfio_t* out);
.ft 1
.fi
This function compresses the entire
.L PZ_WRITE
.I pzip
stream
.L pz
to the output
.I sfio
stream
.LR out .
It returns
.L 0
on success and
.L -1
on error.

.PP
.nf
.ft 5
int        pzinflate(Pz_t* pz, Sfio_t* out);
.ft 1
.fi
This function decompresses the entire
.L PZ_READ
.I pzip
stream
.L pz
to the output
.I sfio
stream
.LR out .
It returns
.L 0
on success and
.L -1
on error.

.PP
.nf
.ft 5
int        pzgethdr(Pz_t* pz);
.ft 1
.fi
This function reads the header from the
.L PZ_READ
.I pzip
stream
.L pz
and fills in the appropriate fields in
.LR pz .
It returns
.L 0
on success and
.L -1
on error.

.PP
.nf
.ft 5
int        pzputhdr(Pz_t* pz, Sfio_t* out);
.ft 1
.fi
This function writes the header from the
.L PZ_WRITE
.I pzip
stream
.L pz
to the
.I sfio
output stream
.LR out .
It returns
.L 0
on success and
.L -1
on error.

.PP
.Ss "STREAM INFORMATION"
.PP
.nf
.ft 5
Pzpart_t*  pzgetpart(Pz_t* pz, const char* name);
.ft 1
.fi
This function returns a partition given its name.
0 is returned if the partition is not found.
The default partition name is the empty string ("").
.PP
.nf
.ft 5
Pzpart_t*  pzsetpart(Pz_t* pz, Pzpart_t* part);
.ft 1
.fi
This function sets the current active partition to part.
The previous active partition is returned.
The current active partition is initialized to the default parition ("").
.PP
.nf
.ft 5
void       pzdump(Pz_t* pz, Sfio_t* out);
.ft 1
.fi
This function writes the header header information from the
.I pzip
stream
.L pz
to the
.I sfio
output stream
.L out
in partition file format.
A file containing this information is suitable for the
.L Pzdisc_t.partition
field.
It returns
.L 0
on success and
.L -1
on error.

.PP
.Ss "SFIO DECOMPRESS DISCIPLINE"
.PP
.nf
.ft 5
int        sfdcpzip(Sfio_t* sp, const char* options, int flags);
.ft 1
.fi
This function pushes a
.I pzip
decompress
.I sfio
discipline on the
.I sfio
stream
.LR sp
by calling
.IR pzopen ()
with
.L PZ_READ|PZ_STREAM
on
.L sp
and setting
.L Pzdisc_t.options
to
.LR options .
Because of the extra information involved,
.I pzip
compression is not supported by the discipline.
This is better handled by the
.IR pzip (1)
command interface.
This means that
.L sp
must be an
.L SF_READ
.I sfio
stream.
.IR sfdcpzip ()
stacks the
.I gzip
decompress discipline
.IR sfdcgzip ()
on
.L sp
if necessary.
0 is returned on success.

.SH AUTHOR
Glenn Fowler, gsf@research.att.com.


src/cmd/pzip/pzip.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1998-2011 AT&T Intellectual Property          *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 1.0                  *
*                    by AT&T Intellectual Property                     *
*                                                                      *
*                A copy of the License is available at                 *
*          http://www.eclipse.org/org/documents/epl-v10.html           *
*         (with md5 checksum b35adb5213ca9657e911e9befb180842)         *
*                                                                      *
*              Information and Software Systems Research               *
*                            AT&T Research                             *
*                           Florham Park NJ                            *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                                                                      *
***********************************************************************/
#pragma prototyped

/*
 * partitioned fixed record zip
 */

static const char usage[] =
"[-?\n@(#)$Id: pzip (AT&T Research) 2003-07-17 $\n]"
USAGE_LICENSE
"[+NAME?\f?\f - fixed record partition compress/decompress]"
"[+DESCRIPTION?\b\f?\f\b compresses and decompresses data files of fixed"
"	length rows (records) and columns (fields). It performs better than"
"	\bgzip\b(1) in space/time on data that has many (typically > 50%)"
"	columns that change at a low rate (columns with a low rate of change"
"	are low frequency; columns with a high rate of change are high"
"	frequency).]"
"[+?The \apzip\a compress format is itself \agzipped\a; decompressed data"
"	is reorganized according to the user-specified \apartition\a file"
"	(see the \b--partition\b option below) before being passed to"
"	\agzip\a. Low frequency columns are difference encoded and high"
"	frequency column groups are transposed to column-major order."
"	The \agzip\a tables are flushed between each column partition group."
"	This has a positive space/time effect on the \agzip\a string match"
"	and huffman tables.]"
"[+?If a \apartition\a file is specified then \apzip\a compresses the input"
"	\afile\a to the standard output, otherwise \apzip\a decompresses"
"	the input \afile\a to the standard ouput. If \afile\a is omitted"
"	then the standard input is used. If the standard input is a tty"
"	then \b/dev/null\b is silently used.]"
"[+?\afile\a may be \apzip\a compressed, \agzip\a compressed, or raw."
"	\apzip\a files self-identify; the row size and partition can be"
"	determined from the \apzip\a header. For \agzip\a and raw data,"
"	the following are done to determine the row size:]{"
"		[+(1)?The row size is taken from the \b--row\b option"
"			if specified.]"
"		[+(2)?If the \b--partition\b option is specified then the"
"			row size is taken from the \apartition\a file.]"
"		[+(3)?If the data is newline-terminated and if it contains"
"			at least two lines and if the first two data lines"
"			have the same length then that length is taken"
"			to be the row size.]"
"		[+(4)?Otherwise the row size cannot be determined and"
"			\apzip\a exits with a diagnostic.]"
"}"

"[a:append?Sets the \bPZ_APPEND\b flag that may be used by some disciplines.]"
"[b:bzip?Use \bbzip\b(1) compression instead of the default \bgzip\b(1)."
"	\abzip\a is not fully supported, pending further investigation.]"
"[c:comment?Place \acomment\a in the output \apzip\a file header when"
"	compressing. The comment is listed by the \b--header\b option.]:"
"		[comment]"
"[x:crc?Enable \agzip\a crc32 cyclic redundancy checking for decompress."
"	On some systems this can double the execution wall time."
"	Most data corruption errors are still caught even with \bnocrc\b.]"
"[d:dump?Enable detailed tracing.]"
"[B:bufsize?Set the output buffer size to \asize\a -- for debugging.]#[size]"
"[D:debug?Set the debug trace level to \alevel\a. Higher levels produce"
"	more output.]#[level]"
"[O:dio?Push the \bsfdcdio\b(3) direct io discipline on the input streams."
"	Silently ignored on systems that do not support direct io.]"
"[G!:gzip?\b--nogzip\b disables \agzip\a compression. Most often used for"
"	conversion or debugging.]"
"[h:header?List header information on the input \apzip\a file and exit."
"	This output is compatible with the \b--partition\b file format.]"
"[l:library?Loads the dll \alibrary\a via the \apzlib\a() call."
"	\alibrary\a must contain the function"
"	\bint pz_init(Pz_t* pz, Pzdisc_t* disc)\b"
"	which is called during \apzip\a stream initialization. \bpz_init\b"
"	allows run time modification to \adisc\a: most often it supplies"
"	alternate discipline functions. Runtime libraries may interpret"
"	options specific to the library; library usage and description"
"	will be appended to online help output if the help options"
"	appear after the \b--library\b option. Runtime libraries may"
"	provide additional diagnostics and tracing when \b--summary\b,"
"	\b--verbose\b or \b--dump\b are specified. In general, runtime"
"	libraries are not needed for decompression. The \b--header\b"
"	option lists the runtime libraries used to compress the input"
"	file.]:[library]"
"[z:lzw?Use \bcompress\b(1) lzw compression instead of the default \bgzip\b(1)."
"	\alzw\a is not fully supported, pending further investigation.]"
"[o:override?Override the column partition. Currently only fixed value"
"	columns may be specified. The syntax is"
"	\abegin\a[-\aend\a]]='\avalue\a' where \abegin\a is the beginning"
"	column offset (starting at 0), \aend\a is the ending column offset"
"	for an inclusive range, and \avalue\a is the fixed column value."
"	Uncompress time is improved when high frequency columns are given"
"	fixed values (see the \b--partition\b option).]:[begin[-end]]=value]"
"[p:partition?\afile\a specifies the data row size and the high frequency"
"	column partition groups and permutation. \afile\a may contain URL-like"
"	components: \apath\a\b?name=\b\apart\a or \apath\a\b#\b\apart\a"
"	reads the partition file \apath\a and uses the partition named"
"	\apart\a. Other options may be set by separating each with , or space."
"	The partition file is a sequence of lines. Comments start with # and"
"	continue to the end of the line. The first non-comment line specifies"
"	the optional name string in \"...\". The next non-comment line"
"	specifies the row size. The remaining lines operate on column offset"
"	ranges of the form: \abegin\a[-\aend\a]] where \abegin\a is the"
"	beginning column offset (starting at 0), and \aend\a is the ending"
"	column offset for an inclusive range. The file name \b//\b or"
"	\b/gzip/\b disables \bpzip\b partitioning and applies only"
"	\bgzip\b compression. The operators are:]:[file]{"
"		[+range [...]]?places all columns in the specified \arange\a"
"			list in the same high frequency partition group."
"			Each high frequency partition group is processed as"
"			a separate block by the underlying compressor"
"			(\bgzip\b(1) by default).]"
"		[+range='value'?specifies that each column in \arange\a"
"			has the fixed character value \avalue\a. C-style"
"			character escapes are valid for \avalue\a.]"
"}"
"[Z:push?Push the \bsfdcpzip\b(3) io discipline rather than direct library"
"	calls. Used for debugging and performance testing.]"
"[P!:pzip?\b--nopzip\b disables \apzip\a compression. Most often used for"
"	conversion or debugging.]"
"[Q:regress?Generate output for regression testing, such that identical"
"	invocations with identical input files will generate the same output.]"
"[r:row?Specifies the input row size (number of byte columns) for data"
"	that does not self-identify.]#[row-size]"
"[S:split?Instead of compressing, the input split discipline, which must be"
"	specified by a subsequent \b--library\b option, splits the input"
"	data into files named \aid\a. \aid\a is determined by the split"
"	discipline \bnamef\b function. The optional \apattern\a is a \bksh\b(1)"
"	file match pattern that limits the split to \aid\a's matching"
"	\apattern\a (e.g., \b--split='1234|98765'\b.) If \b--append\b is also"
"	specified then the data is appened to any pre-existing \aid\a files;"
"	otherwise each file is truncated when the first record containing"
"	\aid\a data is read. If there are no records with \aid\a data then"
"	the \aid\a file is not modified. \b--split\b should be used in a"
"	separate directory, and the directory should be cleared when"
"	\b--append\b is not specified to avoid mixing inconsistent"
"	data. No records will be written to a split file with size"
"	>= \b--window\b bytes.]:?[pattern]"
"[s:summary?Enable summary tracing to the standard error. Runtime libraries"
"	may add addtional information to the default \bpzip\b(3) library"
"	summary output. Compression summary includes the compression rate,"
"	bytes per record, and compression wall time. This option also"
"	enables split discipline warnings about \aid\a partitions that"
"	should be generated by \bpin\b(1) and added to the partition"
"	file to improve compression. The \bpin\b(1) output, with an additional"
"	\"\aid\a\" line manually prepended, can then be appended to an existing"
"	partition file.]"
"[T:test?Enable \bpzip\b(3) implementation-specific tests and tracing.]#"
"		[mask]"
"[v:verbose?Enable intermediate tracing.]"
"[w:window?Each chunk of \awindow\a bytes is compressed separately. The"
"	window size may be silently decreased to accomodate an integral"
"	number of complete rows.]#[window-size:=4M]"
"[W:write-test?Loop on \asfread\a()/\bpzwrite\b(3) in chunks of \agroup\a"
"	records rather than a single \apzdeflate\a() call for compression."
"	Used for debugging and performance testing.]#[group]"
"[X:prefix?Uncompressed data contains a prefix that is defined by \acount\a"
"	and an optional \aterminator\a. This data is preserved but is not"
"	\bpzip\b compressed. If \acount\a is \b0\b on uncompress then the"
"	header is not copied to the output. \aterminator\a may be one"
"	of:]:[count[*terminator]]]{"
"		[+\aomitted\a?\acount\a bytes.]"
"		[+L?\acount\a \bnewline\b terminated records.]"
"		[+'\achar\a'?\acount\a \achar\a terminated records.]"
"}"

"\n"
"\nfile\n"
"\n"
"[+SEE ALSO?\bbzip\b(1), \bgzip\b(1), \bpin\b(1), \bpop\b(1), \bpzip\b(3)]"
"[+BUGS?\bpzip\b decompress currently fails if the standard input is a pipe."
"	This will be addressed in a future release.]"
;

#include <ast.h>
#include <error.h>
#include <pzip.h>
#include <sfdcbzip.h>

typedef int (*Method_f)(Sfio_t*, int);

int
main(int argc, char** argv)
{
	register Pz_t*	pz;
	Pzdisc_t	disc;
	Sfio_t*		dp;
	char*		s;

	Method_f	method = 0;
	ssize_t		bufsize = 0;
	int		push = 0;
	int		testwrite = 0;
	unsigned long	flags = PZ_READ|PZ_FORCE;

	if (s = strrchr(*argv, '/'))
		s++;
	else
		s = *argv;
	error_info.id = s;
	memset(&disc, 0, sizeof(disc));
	disc.version = PZ_VERSION;
	disc.errorf = errorf;
	if (!(dp = sfstropen()))
		error(ERROR_SYSTEM|3, "out of space");
	for (;;)
	{
		switch (optget(argv, usage))
		{
		case 'a':
			flags |= PZ_APPEND;
			continue;
		case 'b':
			method = sfdcbzip;
			continue;
		case 'c':
			disc.comment = opt_info.arg;
			continue;
		case 'd':
			flags |= PZ_DUMP;
			continue;
		case 'h':
			flags &= ~(PZ_READ|PZ_WRITE|PZ_FORCE);
			flags |= PZ_STAT|PZ_DUMP;
			continue;
		case 'l':
			sfprintf(dp, "library=%s\n", opt_info.arg);
			continue;
		case 'o':
			sfputr(dp, opt_info.arg, '\n');
			continue;
		case 'p':
			disc.partition = opt_info.arg;
			flags &= ~(PZ_READ|PZ_FORCE);
			flags |= PZ_WRITE;
			continue;
		case 'r':
			sfprintf(dp, "row=%d\n", opt_info.num);
			continue;
		case 's':
			flags |= PZ_SUMMARY;
			continue;
		case 'S':
			flags |= PZ_SPLIT;
			sfprintf(dp, "split%s%s\n", opt_info.arg ? "=" : "", opt_info.arg ? opt_info.arg : "");
			continue;
		case 'v':
			flags |= PZ_VERBOSE;
			continue;
		case 'w':
			disc.window = opt_info.num;
			continue;
		case 'x':
			flags |= PZ_CRC;
			continue;
		case 'z':
			method = sfdclzw;
			continue;
		case 'B':
			bufsize = opt_info.num;
			continue;
		case 'D':
			error_info.trace = -opt_info.num;
			continue;
		case 'O':
			flags |= PZ_DIO;
			continue;
		case 'G':
			flags |= PZ_NOGZIP;
			continue;
		case 'P':
			flags |= PZ_NOPZIP;
			continue;
		case 'Q':
			sfprintf(dp, "regress\n");
			continue;
		case 'T':
			sfprintf(dp, "test=%s\n", opt_info.arg);
			continue;
		case 'W':
			testwrite = opt_info.num;
			continue;
		case 'X':
			sfprintf(dp, "prefix=%s\n", opt_info.arg);
			continue;
		case 'Z':
			push = 1;
			continue;
		case '?':
			if (opt_info.name[0] == '-' && opt_info.name[1] == '-')
				sfputr(dp, opt_info.name + 2, '\n');
			else
				sfputr(dp, "??short", '\n');
			continue;
		case ':':
			if (!opt_info.option[0] && opt_info.name[0] == opt_info.name[1] || opt_info.option[0] == '-' && opt_info.option[1] == '?')
				sfputr(dp, &argv[opt_info.index - 1][2], '\n');
			else
				error(2, "%s", opt_info.arg);
			continue;
		}
		break;
	}
	argv += opt_info.index;
	if (error_info.errors || *argv && *(argv + 1))
		error(ERROR_USAGE|4, "%s", optusage(NiL));
	if (sftell(dp) && !(disc.options = strdup(sfstruse(dp))))
		error(ERROR_SYSTEM|3, "out of space");
	sfclose(dp);
	if (flags & PZ_SPLIT)
	{
		flags &= ~PZ_WRITE;
		flags |= PZ_READ;
	}
	if (push)
	{
		if (*argv)
			error(3, "%s: file argument not expected for sfdcpzip discipline test", *argv);
		if (sfdcpzip(sfstdin, NiL, flags, &disc) < 0)
			error(3, "sfdcpzip discipline push error");
		if (sfmove(sfstdin, sfstdout, SF_UNBOUND, -1) < 0 || sfclose(sfstdout))
			error(ERROR_SYSTEM|3, "sfdcpzip io error");
		return 0;
	}
	if (method)
	{
		flags |= PZ_NOGZIP;
		if ((*method)((flags & PZ_WRITE) ? sfstdout : sfstdin, 0) < 0)
			error(3, "compression method discipline push error");
	}
	if (bufsize)
	{
		sfset(sfstdout, SF_SHARE|SF_LINE, 0);
		sfsetbuf(sfstdout, NiL, bufsize);
	}
	if (pz = pzopen(&disc, *argv, flags))
	{
		if (testwrite && (flags & PZ_WRITE))
		{
			unsigned char*	buf;
			size_t		n;
			ssize_t		r;

			n = pz->part->row * testwrite;
			error(1, "pzwrite test %d bytes per chunk", n);
			if (!(buf = newof(0, unsigned char, n, 0)))
				error(ERROR_SYSTEM|3, "out of space [buf]");
			while ((r = sfread(pz->io, buf, n)) > 0)
				if (pzwrite(pz, sfstdout, buf, r) != r)
					return 1;
			if (r < 0)
				error(ERROR_SYSTEM|3, "%s: read error", pz->path);
		}
		return ((flags & PZ_WRITE) && pzdeflate(pz, sfstdout) || (flags & PZ_READ) && pzinflate(pz, sfstdout)) || pzclose(pz) || error_info.errors;
	}
	return 1;
}
